source/shadow-tasks/shadow_device_task.c

```diff
@@ -0,0 +1,815 @@
+                                         MQTTAgentReturnInfo_t * pxReturnInfo );
+
+/**
+ * @brief The callback to execute when there is an incoming publish on the
+ * topic for delta updates. It verifies the document and sets the
+ * powerOn state accordingly.
+ *
+ * @param[in] pvIncomingPublishCallbackContext Context of the initial command.
+ * @param[in] pxPublishInfo Deserialized publish.
+ */
+static void prvIncomingPublishUpdateDeltaCallback( void * pxSubscriptionContext,
+                                                   MQTTPublishInfo_t * pxPublishInfo );
+
+/**
+ * @brief The callback to execute when there is an incoming publish on the
+ * topic for accepted requests. It verifies the document is valid and is being waited on.
+ * If so it updates the last reported state and notifies the task to inform completion
+ * of the update request.
+ *
+ * @param[in] pvIncomingPublishCallbackContext Context of the initial command.
+ * @param[in] pxPublishInfo Deserialized publish.
+ */
+static void prvIncomingPublishUpdateAcceptedCallback( void * pxSubscriptionContext,
+                                                      MQTTPublishInfo_t * pxPublishInfo );
+
+
+/**
+ * @brief The callback to execute when there is an incoming publish on the
+ * topic for rejected requests. It verifies the document is valid and is being waited on.
+ * If so it notifies the task to inform completion of the update request.
+ *
+ * @param[in] pvIncomingPublishCallbackContext Context of the initial command.
+ * @param[in] pxPublishInfo Deserialized publish.
+ */
+static void prvIncomingPublishUpdateRejectedCallback( void * pxSubscriptionContext,
+                                                      MQTTPublishInfo_t * pxPublishInfo );
+
+/**
+ * @brief Entry point of shadow demo.
+ *
+ * This main function demonstrates how to use the macros provided by the
+ * Device Shadow library to assemble strings for the MQTT topics defined
+ * by AWS IoT Device Shadow. It uses these macros for topics to subscribe
+ * to:
+ * - SHADOW_TOPIC_STRING_UPDATE_DELTA for "$aws/things/thingName/shadow/update/delta"
+ * - SHADOW_TOPIC_STRING_UPDATE_ACCEPTED for "$aws/things/thingName/shadow/update/accepted"
+ * - SHADOW_TOPIC_STRING_UPDATE_REJECTED for "$aws/things/thingName/shadow/update/rejected"
+ *
+ * It also uses these macros for topics to publish to:
+ * - SHADOW_TOPIC_STIRNG_DELETE for "$aws/things/thingName/shadow/delete"
+ * - SHADOW_TOPIC_STRING_UPDATE for "$aws/things/thingName/shadow/update"
+ */
+void vShadowDeviceTask( void * pvParameters );
+
+/*-----------------------------------------------------------*/
+
+static bool prvSubscribeToShadowUpdateTopics( void )
+{
+    bool xReturnStatus = true;
+    MQTTStatus_t xStatus;
+    uint32_t ulNotificationValue;
+    CommandInfo_t xCommandParams = { 0 };
+
+    /* These must persist until the command is processed. */
+    static MQTTAgentSubscribeArgs_t xSubscribeArgs;
+    static MQTTSubscribeInfo_t xSubscribeInfo[ 3 ];
```

Comment:
Do these need to be static given that we perform a wait and assert on line 320? Not making the command context static would mean line 277 could be deleted